lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,704 @@
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// The issuer of the offer, possibly beginning with `user@domain` or `domain`. Intended to be
+	/// displayed to the user but with the caveat that it has not been verified in any way.
+	pub fn issuer(&self) -> Option<PrintableString> {
+		self.contents.issuer.as_ref().map(|issuer| PrintableString(issuer.as_str()))
+	}
+
+	/// Paths to the recipient originating from publicly reachable nodes. Blinded paths provide
+	/// recipient privacy by obfuscating its node id.
+	pub fn paths(&self) -> &[BlindedPath] {
+		self.contents.paths.as_ref().map(|paths| paths.as_slice()).unwrap_or(&[])
+	}
+
+	/// The quantity of items supported.
+	pub fn supported_quantity(&self) -> Quantity {
+		self.contents.supported_quantity()
+	}
+
+	/// The public key used by the recipient to sign invoices.
+	pub fn signing_pubkey(&self) -> PublicKey {
+		self.contents.signing_pubkey.unwrap()
+	}
+
+	#[cfg(test)]
+	fn as_tlv_stream(&self) -> OfferTlvStreamRef {
+		self.contents.as_tlv_stream()
+	}
+}
+
+impl OfferContents {
+	pub fn amount_msats(&self) -> u64 {
+		self.amount.as_ref().map(Amount::as_msats).unwrap_or(0)
+	}
+
+	pub fn supported_quantity(&self) -> Quantity {
+		self.supported_quantity
+	}
+
+	fn as_tlv_stream(&self) -> OfferTlvStreamRef {
+		let (currency, amount) = match &self.amount {
+			None => (None, None),
+			Some(Amount::Bitcoin { amount_msats }) => (None, Some(*amount_msats)),
+			Some(Amount::Currency { iso4217_code, amount }) => (
+				Some(iso4217_code), Some(*amount)
+			),
+		};
+
+		let features = {
+			if self.features == OfferFeatures::empty() { None } else { Some(&self.features) }
+		};
+
+		OfferTlvStreamRef {
+			chains: self.chains.as_ref(),
+			metadata: self.metadata.as_ref(),
+			currency,
+			amount,
+			description: Some(&self.description),
+			features,
+			absolute_expiry: self.absolute_expiry.map(|duration| duration.as_secs()),
+			paths: self.paths.as_ref(),
+			issuer: self.issuer.as_ref(),
+			quantity_max: self.supported_quantity.to_tlv_record(),
+			node_id: self.signing_pubkey.as_ref(),
+		}
+	}
+}
+
+impl Writeable for OfferContents {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		self.as_tlv_stream().write(writer)
+	}
+}
+
+/// The minimum amount required for an item in an [`Offer`], denominated in either bitcoin or
+/// another currency.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Amount {
+	/// An amount of bitcoin.
+	Bitcoin {
+		/// The amount in millisatoshi.
+		amount_msats: u64,
+	},
+	/// An amount of currency specified using ISO 4712.
+	Currency {
+		/// The currency that the amount is denominated in.
+		iso4217_code: CurrencyCode,
+		/// The amount in the currency unit adjusted by the ISO 4712 exponent (e.g., USD cents).
+		amount: u64,
+	},
+}
+
+impl Amount {
+	/// Returns the amount in millisatoshi.
+	pub fn as_msats(&self) -> u64 {
+		match self {
+			Amount::Currency { .. } => unimplemented!(),
```

Comment:
In `InvoiceRequestBuilder`, we need to check that the amount provided is sufficient for the `Offer` when accounting for both currency and quantity to be accordance with the spec:

```
- MUST specify `invreq_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency` and `invreq_quantity`)
```

Reply:
> So IIUC that means we have to do currency conversion when we receive an invoicerequest and check that the msat amount matches the given-currency amount? And for that you want to have a util on the `Offer` to do the conversion, presumably with some trait for currency conversion?
> 

Was thinking more on the writer side, but yes also true when reading an `invoice_request` from the wire if the `offer` is not denominated in BTC. Would need some sort of type parameter to fetch the exchange rage. I assume you're worried about blocking on network I/O?

> For now, let's just leave it all our, skip the conversion function, and reject invoicerequests with a currency other than bitcoin?

Well, an `invoice_request` is always in terms of msats, so there's no means to reject one. It's only the `offer` that may specify a currency. Everything is in msats after that.

> That avoids adding broken code here, and we can figure out how/if we want to support currency conversion in the onion-message-handling pipeline later.

We still need it on the `InvoiceRequestBuilder` side after scanning an `Offer`. Even if we don't support writing a non-BTC `Offer`, we still need to be able to request an invoice (in msats) if a user scans one.

Also note that if an offer is meant to be long-lived, denominated in another currency is pretty essential to avoid large fluctuations in the exchange rate.

Reply:
> Yes, building an async trait to fetch the exchange rate in the middle of a sync message process is.... not going to be fun.

Could we do the same as we do for fee rates? Just have a background task that polls the latest exchange rates.

> What I meant is, if the offer is for $20 USD, and we get an invoice_request for $10 in msats, we have to reject that, no?

Correct. Yeah, we're on the same page now. Needed both for a sender requesting an `invoice` from an `offer` denominated in USD and a recipient who publishes said `offer` when checking an `invoice_request`.

> At least there we can force the user to provide it - the request_invoice method can require users pass in the msat value for offers that are in other currencies.
> 

Sure, that's definitely an alternative API. But if we need to check an amount when handling an `invoice_request`, why not be consistent?

> I don't buy this. If you post a permanent offer, it's probably not for a specific amount - you're asking for donations. If you have a specific amount in mind, it's for a specific payer, and probably only valid for some short time period (to pay an order).
> 
> Thinking more about it I'm not sure how much of a problem different currency denomination offers is solving. That horse is out of the barn now so not worth trying to revisit it, but I don't think we should bend over backwards building something clean for that use-case until we see how people are using offers in the wild.

I'd imagine it's useful for recurring payments / subscriptions.

Reply:
> We currently push that complexity on users, which kinda sucks. Further, for exchange rates things may be more complicated - you have to pre-fetch _every_ exchange rate and store that, not just one three values but hundreds. Worse, I'm not sure what the appropriate answer is for a node that doesn't want to query some centralized website and give them a list of all users running lightning - how are we going to do exchange rate lookups for the sample node? We certainly can't query some centralized website, and we don't currently directly support SOCKS to pass it over tor, nor do I know that we want to require Tor to run it.
> 
> To make matters worse, most financial data APIs are paid, or at least paid over some rate threshold. A cursory glance at exchange rate data seems to indicate that's mostly true here too.
> 

Fair points. Though, FWIW, a user only needs to poll for the currency in which they denominated offers in. For paying offers, the implementation can fetch on demand or return an error if unsupported.

> Mmm, recurring payments is a good point. IIRC that's been pushed from the v1 spec for now?
> 

Correct.

> Right, my question is how to make this simple for users. On the receiving side, you only need to care if you ever provided an offer with a non-BTC currency, on the sending side you always need to care. I don't really have an answer here, mostly because I'm pretty confident all possible answers really really suck. I think it may be totally reasonable for a node to simply not support sending to non-BTC-denominated offers, sadly.
> 

If `InvoiceRequestBuilder::amount_msat` (or the builder itself) is parameterized by the exchange trait, then we could define a default implementation that errors, I suppose, if `Offer::amount` is a `Currency`.

> In any case, as for the current PR, I'd really rather not commit code that panics and can't really be fixed, and at least for the offer-issuance side we can trivially support only-btc-denominated offers for now, which seems like a simple way to go.

So essentially have `InvoiceRequestBuilder` error if attempting to pay a non-BTC-denominated offer?

Reply:
Changed `OfferBuilder::amount` to `OfferBuilder::amount_msats` and updated dependent PRs accordingly.